pubpub/pubpub
# Bh/citations oauth

## Issue(s) Resolved
#2414
#2413 

+ Introduces `Integration` and `IntegrationData[auth schema name]` models which store auth data serverside.
+ Adds new route for authentication at `/auth/zotero`
+ Adds 'activate'/'delete' integration buttons to the privacy settings at `/legal/privacy`
+ introduces a js zotero client, which will be developed further in #2408

## Test Plan
in order to test this in dev, you will need to add `process.env.ZOTERO_CLIENT_KEY` and `process.env.ZOTER_CLIENT_SECRET` variables to your `config.js` (I saved these to the onepass secure note for config at the top of the 'global' section), as well as get the login credentials for the knowledge futures zotero test account (saved to the pubpub vault on onepass)

test creation:

1. log in
2. navigate to user privacy settings
3.  click 'activate' for zotero integration
4. proceed through oauth flow, including login to zotero and return to privacy settings
5. back on privacy settings, confirm the zotero integration offers option to delete integration

test deletion:

1. following the above, get to privacy settings
2. click 'delete' for the zotero integration
3. it should immediately switch apparent status in UI
4. refresh page to confirm the integration is deleted

more granular testing available by observing rows in the tables `Integrations` and `IntegrationDataOAuth1` in dev DB.

## Screenshots (if applicable)
![Screenshot from 2023-02-08 22-35-34](https://user-images.githubusercontent.com/14115194/217711515-9583f16c-b4f3-4de6-8714-317d5d6d98b5.png)

## Optional

### Notes/Context/Gotchas
repeat visits to the auth url (i.e. `/auth/zotero` should refresh the authentication token in the `IntegrationDataOAuth1` row

### Supporting Docs
